Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp

```diff
@@ -206,82 +206,99 @@ static void tryCreateNativeImageFromData(std::span<const uint8_t> data, std::opt
+{
+    Vector<RefPtr<ShareableBitmap>> bitmaps;
+    for (auto length : lengths) {
+        if (RefPtr nativeImage = nativeImageFromSVGImage(image, length)) {
+            if (RefPtr bitmap = ShareableBitmap::createFromImageDraw(*nativeImage, DestinationColorSpace::SRGB(), { (int)length, (int)length }, ShareableBitmap::ShouldShrinkToFit::Yes)) {
+                bitmaps.append(WTFMove(bitmap));
+                continue;
+            }
+        }
+        return { };
     }
 
-    if (!CGImageDestinationFinalize(destination.get()))
-        return nullptr;
-
-    return SharedBuffer::create(destinationData.get());
+    return bitmaps;
 }
 
-void createIconDataFromImageData(std::span<const uint8_t> data, std::span<const unsigned> lengths, CompletionHandler<void(RefPtr<SharedBuffer>&&)>&& completionHandler)
+void getBitmapsForIconFromImageData(std::span<const uint8_t> data, std::span<const unsigned> lengths, CompletionHandler<void(Vector<RefPtr<ShareableBitmap>>&&)>&& completionHandler)
 {
     // Supported ICO image sizes by ImageIO.
     constexpr std::array<unsigned, 5> availableLengths { { 16, 32, 48, 128, 256 } };
-    constexpr auto icoUTI = "com.microsoft.ico"_s;
     auto targetLengths = lengths.empty() ? std::span { availableLengths } : lengths;
 
     if (RefPtr nativeImage = tryCreateNativeImageFromBitmapImageData(data, std::nullopt)) {
-        completionHandler(expandNativeImageToData(*nativeImage, icoUTI, targetLengths));
+        completionHandler(bitmapsFromNativeImage(*nativeImage, targetLengths));
         return;
     }
 
-    SVGImage::tryCreateFromData(data, [icoUTI, targetLengthsVector = Vector<unsigned> { targetLengths }, completionHandler = WTFMove(completionHandler)](auto svgImage) mutable {
+    SVGImage::tryCreateFromData(data, [targetLengthsVector = Vector<unsigned> { targetLengths }, completionHandler = WTFMove(completionHandler)](auto svgImage) mutable {
         if (!svgImage) {
-            completionHandler(nullptr);
+            completionHandler({ });
             return;
         }
-        completionHandler(expandSVGImageToData(*svgImage, icoUTI, targetLengthsVector.span()));
+        completionHandler(bitmapsFromSVGImage(*svgImage, targetLengthsVector.span()));
     });
 }
 
+RefPtr<SharedBuffer> createIconDataFromBitmaps(Vector<RefPtr<ShareableBitmap>>&& bitmaps)
+{
+    if (bitmaps.isEmpty())
+        return nullptr;
+
+    constexpr auto icoUTI = "com.microsoft.ico"_s;
+    RetainPtr cfUTI = icoUTI.createCFString();
+    RetainPtr colorSpace = adoptCF(CGColorSpaceCreateWithName(kCGColorSpaceSRGB));
+    RetainPtr destinationData = adoptCF(CFDataCreateMutable(0, 0));
+    RetainPtr destination = adoptCF(CGImageDestinationCreateWithData(destinationData.get(), cfUTI.get(), bitmaps.size(), nullptr));
+
+    for (auto bitmap : bitmaps) {
+        if (!bitmap)
+            return nullptr;
+
+        RetainPtr cgImage = bitmap->makeCGImageCopy();
+        if (!cgImage)
+            return nullptr;
```

Comment:
Because makeCGImageCopy() can return nullptr, should we log some information about this image?
```
        LOG_WITH_STREAM(Images, stream
            << "createIconDataFromBitmaps() failed ShareableBitmap::makeCGImageCopy()"
            << " bitmap size: " << bitmap->size());
```